File: Java/src/ch10/DFS.java
package ch10;

import java.util.HashSet;
import java.util.Stack;

// 深度优先遍历
public class DFS {

    /**
     * 深度优先
     * @param node 出发节点
     * @return
     */
    public static HashSet<Node> dfs(Node node){

        // 做一个是否为空的判断
        if (null == node){
            return null;
        }
        // 定义一个栈
        Stack<Node> stack = new Stack<>();
        // 用一个set来记录已经遍历过的节点
        HashSet<Node> set = new HashSet<>();
        // 进栈
        stack.add(node);
        // 记录
        set.add(node);
        // 打印
        System.out.println(node);
        while (!stack.empty()){
            // 栈，后进先出
            Node cur = stack.pop();
            // 遍历 next
            for (Node iter: cur.nexts){
                // 这个节点没有被遍历过
                if (!set.contains(iter)) {
                    // 把cur压回栈
                    stack.add(cur);
                    // 把当前节点压回栈
                    stack.add(iter);
                    // 记录
                    set.add(iter);
                    // 打印
                    System.out.println(iter);
                    // break，不再循环遍历，从iter出发
                    break;
                }
            }
        }
        return set;
    }

    public static void main(String[] args) {
        Integer[][] m = {{null,1,2},{null,2,1},
                         {null,2,3},{null,3,2},
                         {null,1,4},{null,4,1},
                         {null,2,5},{null,5,2},
                         {null,3,6},{null,6,3},
                         {null,4,5},{null,5,4},
                         {null,5,6},{null,6,5},
                         {null,5,7},{null,7,5},
                         {null,6,8},{null,8,6},
                         {null,7,8},{null,8,7}
                        };
        Graph g = Converter.FromMatrixToGraph(m);
        HashSet<Node> set = dfs(g.nodes.get(1));
        System.out.println(set);
    }
}

File: Python/ch10/DFS.py
from ch10 import Converter


def dfs(node):
    # 做一个是否为空的判断
    if node is None:
        return None
    # 定义一个栈
    # 我们用Python的list
    # append()表示入栈
    # pop()表示出栈
    stack = []
    # 用一个set来记录已经遍历过的节点
    s = set()
    # 进栈
    stack.append(node)
    # 记录
    s.add(node)
    # 打印
    print(node)
    while len(stack) > 0:
        # 栈，后进先出
        cur = stack.pop()
        # 遍历 next
        for iter in cur.nexts:
            # 这个节点没有被遍历过
            if iter not in s:
                # 把cur压回栈
                stack.append(cur)
                # 把当前节点压回栈
                stack.append(iter)
                # 记录
                s.add(iter)
                # 打印
                print(iter)
                # break，不再循环遍历，从iter出发
                break
    return s


if __name__ == '__main__':
    m = [[None, 1, 2], [None, 2, 1],
         [None, 2, 3], [None, 3, 2],
         [None, 1, 4], [None, 4, 1],
         [None, 2, 5], [None, 5, 2],
         [None, 3, 6], [None, 6, 3],
         [None, 4, 5], [None, 5, 4],
         [None, 5, 6], [None, 6, 5],
         [None, 5, 7], [None, 7, 5],
         [None, 6, 8], [None, 8, 6],
         [None, 7, 8], [None, 8, 7]]
    g = Converter.from_matrix_to_graph(m)
    s = dfs(g.nodes.get(1))
    print('-' * 20)
    for iter in s:
        print(iter)
